### PR TITLE
Drivetrain

### PR DESCRIPTION
**Pull request by SolidityContract**
Added Drivetrain 
Added PIDTurn and PIDDrive Commands
Updated Constants

### CODE REVIEW

**Review by akashsarada — CHANGES_REQUESTED** [review]
Couple things need to be changed:

1. Make the DrivetrainConstants Class a interface
2. Add follower motors to clean up the code (each side of the drivetrain follows the front motor)
3.  Remove the pigeon from the resetSensors method and rename to resetEncoders
4. Rename PID Controllers to drivePID and turnPID

**Review by akashsarada — APPROVED** [review]
LGTR (literally)

**Review by DarkMG1 — APPROVED** [review]
Good!

**Review by akashsarada — CHANGES_REQUESTED** [review]
Constants file needs to be worked on.

1. Having a final variable in an interface is redundant, can we remove all the final tags?
3. Can we replace the k{Name} nomeclature with a {UPPER_CASE} nomeclature for all the variable names other than those effecting PID values?
4. Can we remove the ID tags for all the variables associating with the CAN_ID, and just group them together?
5. Can we add grouping comments?

**Review by akashsarada — COMMENTED** [review]
Need to rework constants

TODO: @akashsarada

**Review by akashsarada — APPROVED** [review]
Good to Go!

**Review by kelmalon — CHANGES_REQUESTED** [review]
Please edit your java docs in DriveTrain, and add Accelerometer code. See Javadocs Google Doc

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 1]
```diff
@@ -102,15 +103,15 @@ public void tankDrive(double left, double right, double nerf) {
      * @return encoder position based on encoder values
      */
     public double getPosition() {
-        return (frontRight.getEncoder().getPosition() + frontLeft.getEncoder().getPosition()) / (2 * DriveConstants.GEARBOX_SCALE);
+        return (frontRight.getEncoder().getPosition() + frontLeft.getEncoder().getPosition()) / (2 * Constants.DriveTrain.GEARBOX_SCALE);
     }
 
     public double getLeftPosition() {
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:109]: Add javadocs, and return the average of the variables set within periodic. Remove Gearbox scale, we want the actual averaged encoder values to compare if any issues arise.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 2]
```diff
@@ -102,15 +103,15 @@ public void tankDrive(double left, double right, double nerf) {
...
-        return frontLeft.getEncoder().getPosition() / DriveConstants.GEARBOX_SCALE;
+        return frontLeft.getEncoder().getPosition() / Constants.DriveTrain.GEARBOX_SCALE;
     }
 
     public double getRightPosition() {
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:113]: Add javadocs, and return the average of the variables set within periodic. Remove Gearbox scale, we want the actual averaged encoder values to compare if any issues arise.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 3]
```diff
@@ -102,15 +103,15 @@ public void tankDrive(double left, double right, double nerf) {
...
-        return frontRight.getEncoder().getPosition() / DriveConstants.GEARBOX_SCALE;
+        return frontRight.getEncoder().getPosition() / Constants.DriveTrain.GEARBOX_SCALE;
     }
 
     /**
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:117]: "Resets the all DriveTrain encoders to a position of 0." Remove return tag

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 4]
```diff
@@ -132,26 +133,26 @@ public void resetEncoders() {
      * @return the angle to 0
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:133]: add "...of the gyroscope inside the pigeon to 0." Remove return tag

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 5]
```diff
@@ -132,26 +133,26 @@ public void resetEncoders() {
...
-//        while (Math.abs(output) > 180) {
-//            if (output < 0) {
-//                output += 360;
-//            } else {
-//                output -= 360;
-//            }
-//        }
-//        return output;
-//    }
+    public double getAngle() {
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:145]: Why are we only returning a value between -180 and 180? Robot rotates 360 degrees....

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 6]
```diff
@@ -132,26 +133,26 @@ public void resetEncoders() {
...
+                output += 360;
+            } else {
+                output -= 360;
+            }
+        }
+        return output;
+    }
 
 
     /**
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:158]: Something like "Uses PIDController to drive the robot a certain distance based on the average of the left and right encoder values."

Comment (kelmalon) [reply src/main/java/org/troyargonauts/subsystems/DriveTrain.java:158]: Change to 1 Drivetrain PID not 2; Specify UNITS of distance, Tag name is not part of description, will display as "that moved robot to setpoint"

**Review by kelmalon — CHANGES_REQUESTED** [review]
Please reach out with any questions

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java (hunk 3 above)
Comment (akashsarada) [reply src/main/java/org/troyargonauts/subsystems/DriveTrain.java:117]: This function does not reset the drivetrain encoders.

Remove gearbox scale.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java (hunk 4 above)
Comment (akashsarada) [reply src/main/java/org/troyargonauts/subsystems/DriveTrain.java:133]: "Resets the all DriveTrain encoders to a position of 0." - @kelmalon

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java (hunk 3 above)
Comment (kelmalon) [reply src/main/java/org/troyargonauts/subsystems/DriveTrain.java:117]: Unsure how that happened, oops. That is referring to the reset encoder function that I was unable to comment on I believe.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java (hunk 4 above)
Comment (kelmalon) [reply src/main/java/org/troyargonauts/subsystems/DriveTrain.java:133]: Must've gotten switched ¯\_(ツ)_/¯

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 7]
```diff
@@ -13,19 +13,20 @@
 import org.troyargonauts.Robot;
 
 /**
- * using PID and stating our speed, turn, and nerf we made a code to run our 8 wheel tank drivetrain with 2 motors
+ * Drivetrain uses PID as well as runs with certian speed and turn to reach its desired target
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:16]: "Drivetrain allows control of the robot's drivetrain in an arcade drive (cheesy drive) and tank drive. Includes a distance PID and a turn PID using the Pigeon 2 from CTRE" Something like this. Describe the entire class in 2-3 sentences, not just a feature.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 8]
```diff
@@ -13,19 +13,20 @@
...
-    Pigeon2 pigeon;
+    private final Pigeon2 pigeon;
 
-    PIDController leftPID, rightPID, turnPID;
+    private final PIDController drivePID, turnPID;
 
     /**
-     * Creates a new drivetrain object for the code and states the motors needed for the drivetrain
+     * Constructor for the robot's Drivetrain. Instantiates motor controllers, changes encoder convertion factors and instantiates pid controllers.
+     * Motor controllers are reversed and set to follow other motor controllers.
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:29]: Which side is reversed? Also capitalize PID

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 9]
```diff
@@ -82,41 +81,58 @@ public void periodic() {
...
-     * @param nerf decreases the max speed and amount we want to turn the robot
+     * Sets motors value based on speed and turn parameters. 
+     * Robots speed and turn will be controlled by different joysticks.
+     * Allows robot to move in specified direction with more control.
+     * @param speed speed of robot.
+     * @param turn amount we want to turn.
+     * @param nerf decreases the max speed and amount we want to turn the robot.
      */
     public void cheesyDrive(double speed, double turn, double nerf) {
         frontRight.set(((speed - turn) + 0.0010) * nerf);
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:92]: Add right side correction to Constants.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 10]
```diff
@@ -82,41 +81,58 @@ public void periodic() {
...
+    /** 
+     * Sets left and right motor values based on left and right parameters. 
+     * Robots left and right side will be controlled by different joysticks.
+     * Used for mainly testing and troubleshooting.
+     * @param left speed of the left side of the robot.
+     * @param right speed of the right side of the robot.
+     * @param nerf decreases the max speed and amount we want to turn the robot.
+     */
     public void tankDrive(double left, double right, double nerf) {
         frontRight.set(right * nerf);
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:105]: Add right side correction constant here.

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 11]
```diff
@@ -154,67 +168,42 @@ public double getAngle() {
...
-    }
-
-    
     /** 
-     * Drives certian distance based parameter
-     * @param setpoint distance away we want robot to be
-     * @return PIDCommand that moved robot to setpoint
+     * Uses PIDController to turn the robot a certain angle based on the pigeons yaw
+     * @param angle the angle we want the robot to be at
+     * @return PIDCommand that turns robot to target angle
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:174]: Add "PID Command" before "that turns robot to target angle"

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 12]
```diff
@@ -154,67 +168,42 @@ public double getAngle() {
...
                 output -> cheesyDrive(output, 0, 1),
                 Robot.getDrivetrain()
         );
     }
 
+    /** 
+     * Causes the robot to break.
+     * Resets encoders to 0.
+     */
     public void breakMode() {
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:205]: Refactor to BRAKE mode. Fix spelling in Javadoc.

**Review by kelmalon — CHANGES_REQUESTED** [review]
Just a few more changes! Looking good!

**Review by kelmalon — COMMENTED** [review]
Please see unresolved conversations, those are the only changes that need to be made, then it will be ready to pull into master.

**Review by kelmalon — COMMENTED** [review]
Looks good, don't forget to add Accelerometer Data

**Review by kelmalon — APPROVED** [review]

File: src/main/java/org/troyargonauts/subsystems/DriveTrain.java [hunk 13]
```diff
@@ -154,67 +168,42 @@ public double getAngle() {
...
+                drivePID,
                 () -> getPosition(),
                 setpoint,
                 output -> cheesyDrive(output, 0, 1),
                 Robot.getDrivetrain()
         );
     }
 
+    /** 
+     * Causes the robot to break.
```
Comment (kelmalon) [inline src/main/java/org/troyargonauts/subsystems/DriveTrain.java:218]: BRAKE mode

**Review by kelmalon — CHANGES_REQUESTED** [review]

**Review by kelmalon — COMMENTED** [review]
Looks good! Please make the 1 change left to BrakeMode() Javadoc

**Review by kelmalon — CHANGES_REQUESTED** [review]
REMOVE Pneumatics please

**Review by kelmalon — APPROVED** [review]
Looks perfect, thank you!

**Pull request merged**